notebooks/2.modeling/nbconvert/modeling.py

```diff
@@ -0,0 +1,435 @@
+X = training_df[feat_cols].values
+y_labels = training_df["injury_code"]
+
+# since this is a multi-class problem and in order for precision and recalls to work
+# we need to binarize it to different classes
+# source: https://stackoverflow.com/questions/56090541/how-to-plot-precision-and-recall-of-multiclass-classifier
+n_classes = len(np.unique(y_labels.values))
+y = label_binarize(y_labels, classes=[*range(n_classes)])
+
+# then we can split the data set with are newly binarized labels
+# we made sure to use stratify to ensure proportionality within training and test datasets
+X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=seed, stratify=y)
+
+
+# In[6]:
+
+
+def evaluate(
+    model: BaseEstimator,
+    X: np.ndarray,
+    y: np.ndarray,
+    dataset: str,
+    shuffled: bool,
+    seed: Optional[int] = 0,
+) -> tuple[pd.DataFrame]:
+    """calculates the precision/recall and f1
+
+    Parameters
+    ----------
+    model : BaseEstimator
+        best model
+    X : np.ndarray
+        features
+    y : np.ndarray
+        labels
+    shuffled : bool
+        Flag indicating if the data has been shuffled
+    seed : Optional[int], optional
+        _description_, by default 0
+
+    Returns
+    -------
+    tuple[pd.DataFrame]
+        returns a tuple that contains the f1 scores and precision/recall scores
+        in a dataframe
+    """
+
+    # setting seed
+    np.random.seed(seed)
+
+    # number of classes
+    n_classes = len(np.unique(y, axis=0))
+
+    # making predictions
+    predictions = model.predict(X)
+    probability = model.predict_proba(X)
+
+    # computing and collecting  precision and recall curve
+    precision_recall_scores = []
+    for i in range(n_classes):
+        # precision_recall_curve calculation
+        precision, recall, _ = precision_recall_curve(y[:, i], probability[:, i])
+
+        # iterate all scores and save all data into a list
+        for i in range(len(precision)):
+            precision_recall_scores.append([dataset, shuffled, precision[i], recall[i]])
+
+    # creating scores df
+    precision_recall_scores = pd.DataFrame(
+        precision_recall_scores, columns=["dataset", "shuffled", "precision", "recall"]
+    )
+
+    # Compute F1 score
+    f1_scores = []
+    for i in range(n_classes):
+        y_true = y[:, i]
+        y_pred = predictions[:, i]
+        f1 = f1_score(y_true, y_pred)
+        f1_scores.append([dataset, shuffled, injury_codes["decoder"][str(i)], f1])
+
+    # convert to data frame and display
+    f1_scores = pd.DataFrame(
+        f1_scores, columns=["data_set", "shuffled", "class", "f1_score"]
+    )
+
+    return (precision_recall_scores, f1_scores)
+
+
+# ## Training and Evaluating Multi-class Logistic Model with original dataset split
+#
+
+# In[7]:
+
+
+# train and get the best_model
+best_model = train_multiclass(X_train, y_train, param_grid=param_grid, seed=0)
+
+
+# In[ ]:
```

Comment:
Double checking, does this cell (and the others below) need to be re-run before merging?